### ISSUE TITLE
(Feature Request) (Discussion) New NPC Codes

### ISSUE THREAD

**Issue by P-Star7** [Wohlstand/TheXTech#117]
- [ ] New jumphurt codes (jump, spinjump/yoshijump, kuribojump/tanookijump)
How the player reacts to the three types of jumps when jumping onto the NPC. Each of these three jump types can have a different reaction. All of these get ignored if playerblocktop=1, so player will stand safely on the NPC and not bounce off of them.
The second two sets of codes are synonyms for each other. If both are in a file and have conflicting values, the former label in the list takes precedence.

0 = Harm NPC (death for goomba, shell for koopa, shrink for rex, harmed for boss, etc...)
1 = Super-kill (the effect of a spinjump/kuribojump on a rex. Some NPCs like Bits, Bots, Tinsuit, bosses, keep their usual effects on a super-kill, as we know. These will be retained.)
2 = Harm player
3 = Deflected (like spinjump on spiny, kuribojump on thwomp)
4 = Directional deflect, player is bounced in the direction of the NPC's side that they jumped on (Larry/Ludwig's shell)
5 = Springboard bounce

- [ ] playerblocktop = 2
If playerblocktop = 0, player cannot stand on NPC. If playerblocktop = 1, player can stand on NPC, but jump reactions are used when resolving kuribojump/tanookijump situation. These two are vanilla, and NPCs such as shy guys have playerblocktop = 1 by default. If playerblocktop = 2, player can stand on NPC and kuribojump/tanookijump parameter is ignored, they will always stand on this NPC. This is how some NPCs can be stood on by normal player, but while in shoe/statue, they will still be killed, and gives creator an option to disable this.

- [ ] nohammer/nopinkbombs/noboomerang/noswordbeams/notanooki/nostatue (all synonyms, as above one.)

It does what it implies. This applies to all objects that can break grey bricks (hammers, pink bombs, boomerangs, sword beams, tanooki slide)

- [ ] notail
Immunity to Raccoon/Tanooki tail swipe

- [ ] noblock
Immunity to get bumped by a block from below. This also applies to the POW block's effect that takes place the NPC is on the ground.

- [ ] nobomb
Immunity to explosions caused by SMB3 bob-omb, SMB2 bomb, and SMB2 bob-omb. Does not apply to Peach's bombs, see above for nohammer! This also does not make it immune to throw the NPCs at it, use the code below!

- [ ] nothrow
Immunity to thrown items like NPCs (Key, Mushroom Block, Shy Guys) and vegetables. This does not make it immune to bomb explosions, use the code above!

- [ ] nohands
Immunity to NPCs that can harm the NPC while in the player's hands (Shy Guys, vegetables, thwomp)

- [ ] harmfulinhands
This determines if the NPC can harm other NPCs while the player is carrying it, or ONLY if it is thrown. 0 = like a SMW key, mushroom block 1 = like a shy guy, vegetable, thwomp.

- [ ] hot/ishot
If it can be blocked by ducking as hammer Mario/Luigi/Peach/Toad and in the Podoboo's shoe.

- [ ] projectile/isprojectile
Determines if this NPC will die upon touching Link's shield. Objects like hammers and Snifit balls are in this category, but are not in the above.

- [ ] vspeed/speed2
The vertical speed multiplier of the NPC. hspeed is also now added as a synonym for speed NPC code.

- [ ] nogravity = 2
nogravity = 0, it has gravity. nogravity = 1, no gravity. nogravity = 2, it has no gravity, but attains gravity in two situations - when the POW block dies, or when Luigi/Toad hits it out of the block.

Unknown methods of handling, that may be applied to one of the above codes, or that are niche:
Toothy immunity
Yoshi fireball immunity (nofireball = 2 and 3?)
Billy Gun immunity
Hammer Bros. hammers' when they are in the player's hands
Grinder immunity
Shell immunity
Lava immunity
Tank treads immunity
Transformation into a new enemy upon stomping (SMB3/SMB1 Koopa, Para-Goomba, Para-Koopa, SMW Koopa, Rex, Para-Galoomba)

- [ ] Expansion of default NPCs' handling to 0 values
For example, if giving a coin nogravity=0, it does not work, a shell grabside = 0, or mega mole nofireball = 0, it does not work. The old 1.3 behavior is handled by compat.ini. Shall I make this a separate issue?

**Comment by P-Star7** [comment]
Added Yoshi's mouth codes and tail/block bump codes.

**Comment by LoveBodhi** [comment]
My idea is `algorithm` field, which sets the NPC AI.

**Comment by Alucard648** [comment]
- [ ] instantkiller
This NPC, ondealing damage to players or other NPCs, will, instead, instant-kills them. Like giant rolling lava boulder, or Mario Forever poison mushroom.

**Comment by LoveBodhi** [comment]
I suggest "waterjump", whatever hurt by jumping on, like SMB Wonder.

**Comment by Shadowblitz16** [comment]
nofallgrab 
Basically means you can't grab the npc by falling ontop of it while holding b
This would make smb3 style shell physics possible
